std/algebra/twistededwards/bandersnatch/point.go

```diff
@@ -98,41 +108,98 @@ func (p *Point) Double(api frontend.API, p1 *Point, curve EdCurve) *Point {
 	return p
 }
 
+// phi endomorphism √-2 ∈ 𝒪₋₈
+// (x,y) → λ × (x,y) s.t. λ² = -2 mod Order
+func (p *Point) phi(api frontend.API, p1 *Point, curve EdCurve) *Point {
+
+	xy := api.Mul(p1.X, p1.Y)
+	yy := api.Mul(p1.Y, p1.Y)
+	f := api.Sub(1, yy)
+	f = api.Mul(f, curve.endo1)
+	g := api.Add(yy, curve.endo0)
+	g = api.Mul(g, curve.endo0)
+	h := api.Sub(yy, curve.endo0)
+
+	p.X = api.DivUnchecked(f, xy)
+	p.Y = api.DivUnchecked(g, h)
+
+	return p
+}
+
+type glvParams struct {
+	lambda, order big.Int
+	glvBasis      ecc.Lattice
+}
+
+var scalarDecompositionHint = hint.NewStaticHint(func(curve ecc.ID, inputs []*big.Int, res []*big.Int) error {
+	var glv glvParams
+	var init sync.Once
+	init.Do(func() {
+		glv.lambda.SetString("8913659658109529928382530854484400854125314752504019737736543920008458395397", 10)
+		glv.order.SetString("13108968793781547619861935127046491459309155893440570251786403306729687672801", 10)
+		ecc.PrecomputeLattice(&glv.order, &glv.lambda, &glv.glvBasis)
+	})
+
+	// TODO: handle properly negative scalars
+	// lambda.nbits() = scalar_max.nbits() = 253
+	// sp[0] is always negative (as a big.Int)?
+	// thus taking -sp[0] here and negating the point in ScalarMul.
+	sp := ecc.SplitScalar(inputs[0], &glv.glvBasis)
+	res[0].Neg(&(sp[0])) // Set
+	res[1].Set(&(sp[1]))
+
+	// figure out how many times we have overflowed
+	// res[2].Mul(res[1], &lambda).Add(res[2], res[0])
+	res[2].Mul(res[1], &glv.lambda).Sub(res[2], res[0])
+	res[2].Sub(res[2], inputs[0])
+	res[2].Div(res[2], &glv.order)
+
+	return nil
+}, 1, 3)
+
+func init() {
+	hint.Register(scalarDecompositionHint)
+}
+
 // ScalarMul computes the scalar multiplication of a point on a twisted Edwards curve
 // p1: base point (as snark point)
 // curve: parameters of the Edwards curve
 // scal: scalar as a SNARK constraint
 // Standard left to right double and add
 func (p *Point) ScalarMul(api frontend.API, p1 *Point, scalar frontend.Variable, curve EdCurve) *Point {
+	// the hints allow to decompose the scalar s into s1 and s2 such that
+	// s1 + λ * s2 == s mod r,
+	// with λ s.t. λ² = -2 mod Order.
+	sd, err := api.NewHint(scalarDecompositionHint, scalar)
+	if err != nil {
+		// err is non-nil only for invalid number of inputs
+		panic(err)
+	}
 
-	// first unpack the scalar
-	b := api.ToBinary(scalar)
+	s1, s2 := sd[0], sd[1]
 
-	res := Point{}
-	tmp := Point{}
-	A := Point{}
-	B := Point{}
+	// s1 + λ * s2 == s + k*r
+	// api.AssertIsEqual(api.Add(s1, api.Mul(s2, &curve.lambda)), api.Add(scalar, api.Mul(&curve.Order, sd[2])))
+	api.AssertIsEqual(api.Sub(api.Mul(s2, &curve.lambda), s1), api.Add(scalar, api.Mul(&curve.Order, sd[2])))
```

Comment:
Aren't there bounds to check as well? If `o` is the order of bandersnatch's prime group, in Z we have
`s1 + λ * s2 - s - k*o = 0`
`s1, s2 < sqrt(o)`
`o<r`
Here you only check that `s1 + λ * s2 = s + k*o mod r`, so it means that in Z `s1 + λ * s2 - s - k*o = c * r`. From the scalar decomposition we know that `s1, s2 < sqrt(o)<sqrt(r)`, but `λ ` is anywhere between `0` and `o`, so to me nothing guarantees that` c=0` here... I feel that something is missing, linked to the emulation of Z/oZ arithmetic in Z/rZ

Reply:
creating a ticket for this #268